In [1]:
#load the classifying models
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier

import numpy as np
import pandas as pd

In [2]:
df =pd.read_csv('College Admissions.csv')
df.head(10)

admit  gre   gpa  rank
0    No  380  3.61     3
1   Yes  660  3.67     3
2   Yes  800  4.00     1
3   Yes  640  3.19     4
4    No  520  2.93     4
5   Yes  760  3.00     2
6   Yes  560  2.98     1
7    No  400  3.08     2
8   Yes  540  3.39     3
9    No  700  3.92     2

#### 1. Format the admit column to numeric so it can be fit to a classification model. (1 Mark)

In [3]:
df['admit'] = df['admit'].map({'Yes':1, 'No':0})
df

admit  gre   gpa  rank
0        0  380  3.61     3
1        1  660  3.67     3
2        1  800  4.00     1
3        1  640  3.19     4
4        0  520  2.93     4
..     ...  ...   ...   ...
395      0  620  4.00     2
396      0  560  3.04     3
397      0  460  2.63     2
398      0  700  3.65     2
399      0  600  3.89     3

[400 rows x 4 columns]

#### 2. Fit the data to a logistic regression model using scikit learn saving 20% of the data for testing. (1 Mark)



In [4]:
#separate train and test data
x = df[['gre','gpa','rank']]
y = df['admit']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.20, random_state=7)

In [5]:
from sklearn.linear_model import LogisticRegression
#fit a logistic regression model to train data
logisticRegr = LogisticRegression(solver = 'liblinear')

In [6]:
logisticRegr.fit(x_train, y_train)

LogisticRegression(solver='liblinear')

#### 3. Fit the data to a support vector machine model using scikit learn saving 20% of the data for testing. (1 Mark)



In [7]:
from sklearn.svm import SVC 
model = SVC(kernel='linear', C=10)
model.fit(x_train, y_train)

SVC(C=10, kernel='linear')

#### 4. Use scikit learn and the test data to create a confusion matrix for each model. Describe which model preforms better based on the confusion matrix and what you used to determine this. (1 Mark)



##### Logistic regression

In [8]:
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import metrics

In [9]:
# Make predictions on entire test data
predictions = logisticRegr.predict(x_test)

In [10]:
# Use score method to get accuracy of model.
score = logisticRegr.score(x_test, y_test)
print('Accuracy Score for Logistic regression:',score)

Accuracy Score for Logistic regression: 0.725


In [11]:
cm = metrics.confusion_matrix(y_test, predictions)
print(cm)

[[51  2]
 [20  7]]


##### support vector machine

In [12]:
from sklearn import metrics
y_pred=model.predict(x_test)

print('Accuracy Score for SVM:',metrics.accuracy_score(y_test,y_pred))

Accuracy Score for SVM: 0.7


In [13]:
cm_svm = metrics.confusion_matrix(y_test,y_pred)
cm_svm

array([[51,  2],
       [22,  5]], dtype=int64)

#### Description

Describe which model preforms better based on accuracy score both of the model Logistic regression and support vector machine are not meet the criteria of being a good model should be at least 80 % to meet the criteria of a good model, in our case Accuracy Score for Logistic regression (which is 0.725 )is higher than Accuracy Score for SVM(which is 0.7) thats mean that Logistic regression is perform better than SVM.

#### 5. For each of the above, use scikit learn to do a cross validation at 5 folds. Describe which model preforms better based on the cross validation and what you used to determine this. (1 Mark)

In [14]:
#cross validation: n_split - number of folds (number of pieces the data set divided to)
from sklearn.model_selection import KFold
kf = KFold(n_splits= 5)
kf

KFold(n_splits=5, random_state=None, shuffle=False)

In [15]:
for train_index, test_index in kf.split(df):
    print(train_index, test_index)

[ 80  81  82  83  84  85  86  87  88  89  90  91  92  93  94  95  96  97
  98  99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115
 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133
 134 135 136 137 138 139 140 141 142 143 144 145 146 147 148 149 150 151
 152 153 154 155 156 157 158 159 160 161 162 163 164 165 166 167 168 169
 170 171 172 173 174 175 176 177 178 179 180 181 182 183 184 185 186 187
 188 189 190 191 192 193 194 195 196 197 198 199 200 201 202 203 204 205
 206 207 208 209 210 211 212 213 214 215 216 217 218 219 220 221 222 223
 224 225 226 227 228 229 230 231 232 233 234 235 236 237 238 239 240 241
 242 243 244 245 246 247 248 249 250 251 252 253 254 255 256 257 258 259
 260 261 262 263 264 265 266 267 268 269 270 271 272 273 274 275 276 277
 278 279 280 281 282 283 284 285 286 287 288 289 290 291 292 293 294 295
 296 297 298 299 300 301 302 303 304 305 306 307 308 309 310 311 312 313
 314 315 316 317 318 319 320 321 322 323 324 325 32

In [16]:
#using cross validation(CV) with the classification model(logistic regression and SVM) to computing accuracy sccores
#creating vectors to store accuracy at each flod of CV
scores_l =[]
scores_svm = []
from sklearn.model_selection import cross_val_score
#if cv is not specified, default is 5 fold CV
scores_l.append(cross_val_score(LogisticRegression(solver='liblinear'),x,y))
scores_svm.append(cross_val_score(SVC(C=10),x,y))


##### Logistic regression

In [17]:
#accuracy for the linear regression
print('Accuracy the Logistic regression:',scores_l)
print('Accuracy mean:',np.mean(scores_l))


Accuracy the Logistic regression: [array([0.7125, 0.75  , 0.725 , 0.6625, 0.725 ])]
Accuracy mean: 0.7150000000000001


##### support vector machine

In [18]:
#accuracy for the SVM
print('Accuracy for SVM:',scores_svm)
print('Accuracy mean:',np.mean(scores_svm))

Accuracy for SVM: [array([0.6875, 0.6875, 0.6875, 0.675 , 0.675 ])]
Accuracy mean: 0.6824999999999999


### Description:

Based on accuracy score both of the model Logistic regression and support vector machine, the Accuracy Score for Logistic regression (which is 0.7150000000000001 )is higher than Accuracy Score for SVM(which is 0.6824999999999999) thats mean that Logistic regression is perform better than SVM.

#### 6. In your own words, describe a difference between the information you were able to take from cross validation compared to the confusion matrix. (1 Mark)



Cross-validation is a technique that  partitioning a dataset into several subsets, training the model on some of subsets, and then evaluating its performance on the rest, it uses k-fold cross-validation where the data is divided into k equally sized subsets, the model is then trained on k-1 and repeated k times, at last the result averaged.

confusion matrix is a matrix that provides more detailed information about the performance of a classification model, it is shows the number of true positives(TP), false positives(FP), true negatives(TN), and false negatives (FN) for each class to evaluate the performance of a classification model.

#### 7. Choosing two of the four options from the rank column, repeat questions 2 – 5 and compare your results. (3 Marks)

##### How has the accuracy of the model changed when this is accounted for? (1 Mark)

In [19]:
df1 = df[df['rank'].isin([2,3])]

In [20]:
df1.head(10)

admit  gre   gpa  rank
0       0  380  3.61     3
1       1  660  3.67     3
5       1  760  3.00     2
7       0  400  3.08     2
8       1  540  3.39     3
9       0  700  3.92     2
13      0  700  3.08     2
15      0  480  3.44     3
17      0  360  2.56     3
18      0  800  3.75     2

In [21]:
#separate train and test data
x = df1[['gre','gpa','rank']]
y = df1['admit']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.20, random_state=0)

In [22]:
logisticRegr.fit(x_train, y_train)

LogisticRegression(solver='liblinear')

In [23]:
# Make predictions on entire test data
pred = logisticRegr.predict(x_test)

In [24]:
# Use score method to get accuracy of model.
score = logisticRegr.score(x_test, y_test)
print('Accuracy Score:',score)
cm = metrics.confusion_matrix(y_test, pred)
print(cm)

Accuracy Score: 0.6727272727272727
[[37  0]
 [18  0]]


In [25]:
#cross validation: n_split - number of folds (number of pieces the data set divided to)
from sklearn.model_selection import KFold
kf = KFold(n_splits= 5)
kf

KFold(n_splits=5, random_state=None, shuffle=False)

In [26]:
for train_index, test_index in kf.split(df1):
    print(train_index, test_index)

[ 55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108
 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126
 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144
 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162
 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180
 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198
 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215 216
 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233 234
 235 236 237 238 239 240 241 242 243 244 245 246 247 248 249 250 251 252
 253 254 255 256 257 258 259 260 261 262 263 264 265 266 267 268 269 270
 271] [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38

In [27]:
#using cross validation(CV) with the classification model(logistic regression, SVM) to computing accuracy sccores
#creating vectors to store accuracy at each flod of CV
scores_l =[]
scores_svm = []
from sklearn.model_selection import cross_val_score
#if cv is not specified, default is 5 fold CV
scores_l.append(cross_val_score(LogisticRegression(solver='liblinear'),x,y))
scores_svm.append(cross_val_score(SVC(C=10),x,y))


In [28]:
#accuracy for the linear regression
print('Accuracy the Logistic regression:',scores_l)
print('Accuracy mean:',np.mean(scores_l))

Accuracy the Logistic regression: [array([0.69090909, 0.69090909, 0.7037037 , 0.7037037 , 0.72222222])]
Accuracy mean: 0.7022895622895623


In [29]:
#accuracy for the SVM
print('Accuracy for SVM:',scores_svm)
print('Accuracy mean:',np.mean(scores_svm))

Accuracy for SVM: [array([0.69090909, 0.69090909, 0.7037037 , 0.7037037 , 0.7037037 ])]
Accuracy mean: 0.6985858585858586


#### Description:

Based on accuracy score both of the model Logistic regression and support vector machine, the Accuracy Score for Logistic regression (which is 0.7022895622895623 )is higher than Accuracy Score for SVM(which is 0.6985858585858586) thats mean that Logistic regression is perform better than SVM.
Compare to the previous model the accuracy of Logistic regression is higher than this model which mean is preform better.